Apromore-Plugins/plugin-core/portal/api/src/main/java/org/apromore/plugin/portal/BundleWebContentService.java

```diff
@@ -0,0 +1,145 @@
+ *     &lt;beans:property name="include" value="/fancyviewer/.*\.(gif|jpg|png|svg)"/&gt;
+ *     &lt;beans:property name="exclude" value="WEB-INF/.*|.*\.class|.*&#47;"/&gt;
+ *   &lt;/beans:bean&gt;
+ * &lt;/osgi:service&gt;
+ * </pre>
+ */
+public class BundleWebContentService implements WebContentService {
+
+    private static Logger LOGGER = LoggerFactory.getLogger(BundleWebContentService.class);
+
+    /**
+     * The bundle containing the resources.
+     */
+    private final Bundle bundle;
+
+    /**
+     * Paths matching the given pattern will never be exposed by this service.
+     */
+    private Pattern excludePattern = Pattern.compile("WEB-INF/.*|.*\\.class|.*/");
+
+    /**
+     * Paths which don't match the given pattern will never be exposed by this service.
+     */
+    private Pattern includePattern = Pattern.compile(".*");
+
+    /**
+     * @param bundleContext  the context whose resources need to be exposed; must not be <code>null</code>
+     */
+    public BundleWebContentService(BundleContext bundleContext) {
+        bundle = bundleContext.getBundle();
+    }
+
+    public String toString() {
+        return super.toString() + "(" + bundle.getClass().getName() + ")";
+    }
+
+    // Bean properties
+
+    /**
+     * @return a regular expression in the format defined by {@link Pattern} which is used to filter
+     *     which resource paths are exposed as web content
+     */
+    public String getExclude() {
+        return excludePattern.pattern();
+    }
+
+    /**
+     * @param pattern  a regular expression in the format defined by {@link Pattern} which will be used to filter
+     *     which resource paths are exposed as web content
+     */
+    public void setExclude(String pattern) {
+        excludePattern = Pattern.compile(pattern);
+    }
+
+    /**
+     * @return a regular expression in the format defined by {@link Pattern} which is used to filter
+     *     which resource paths are exposed as web content
+     */
+    public String getInclude() {
+        return includePattern.pattern();
+    }
+
+    /**
+     * @param pattern  a regular expression in the format defined by {@link Pattern} which will be used to filter
+     *     which resource paths are exposed as web content
+     */
+    public void setInclude(String pattern) {
+        includePattern = Pattern.compile(pattern);
+    }
+
+    // Implementation of WebContentService
+
+    @Override
+    public boolean hasResource(String path) {
+        return !excludePattern.matcher(path).matches() &&
+            includePattern.matcher(path).matches() &&
+            bundle.getResource(path) != null;
+    }
+
+    @Override
+    public InputStream getResourceAsStream(String path) {
+        assert hasResource(path);
+        try {
+            return bundle.getResource(path).openStream();
+
+        } catch (IOException e) {
+            throw new Error("Unable to read " + path, e);
+        }
+    }
+}
```

Comment:
Reviewing the portal-plugin-api as a whole where this class is located, this getResourceAsStream() method duplicates the getResourceAsStream() method of the PortalPlugin interface. The plugin resource API in this module needs refactoring before it gets more difficult to do: e.g. the API exposed via PortalPlugin, DefaultPortalPlugin and the API exposed via these WebContentService/BundleWebContentService/SimpleWebContentService.

Reply:
Looking at other classes in this module (portal-plugin-api), they need a cleanup because the classes added over time are not cohesive (i.e. serving the same API). 
For example:
- These classes should not be here: MainControllerInterface, SessionTab, PortalSelection. They should be part of the Portal.
- Enum Level is unused.
- FileImporterPlugin: how is it related to the PortalPlugin API? 
- PortalPlugin, DefaultPortalPlugin, WebContentServcie, BundleWebContentService, SimpleWebContentService: the first two classes expose the Plugin API including the plugin resource loading API (so they fit in this module), while the last 3 classes are for resource loading API. However, these two groups are not related to each other, i.e. any extension, association, use relationship?

Again, while modifying and adding more classes in this module, I suggest to do a bit refactoring to clean it before it gets more difficult to do: move classes to the right place, renaming methods, move methods to avoid duplication, etc.

Reply:
Refactoring the Portal APIs is a worthwhile endeavor, but it's somewhat beyond the scope I had in mind for this particular PR.  For now I just want BundleWebContentService to replace PortalPluginResourceServlet and all the WebContentService implementations (SimpleWebContentService, BPMNEditorWebContentService, PDWebContentService...)